api/data/refactoring.py

```diff
@@ -242,3 +245,134 @@ def save_lovd_as_vcf(data, save_to="./lovd.vcf"):
 
             f.write("\t".join(record))
             f.write("\n")
+
+
+def process_population_data(df, pop_data, name, pop_ids, index):
+    for pop_id in pop_ids:
+        df.loc[index, f'{name}_ac_{pop_id}'] = 0
+        df.loc[index, f'{name}_an_{pop_id}'] = 0
+    if isinstance(pop_data, list):
+        for pop in pop_data:
+            variant_id = pop['id']
+            df.loc[index, f'{name}_ac_{variant_id}'] = pop['ac']
+            df.loc[index, f'{name}_an_{variant_id}'] = pop['an']
+
+
+def request_gnomad_api_data(gene_name):
+    """
+    Requests gnomAD API for data about a specific gene containing:
+    - variant_id
+    - cDNA change
+    - protein change
+    - allele frequency
+    - homozygote count
+    - popmax
+    - popmax population
+
+    :param str gene_name: name of gene
+    :param bool to_file: if True, saves data to variants.csv
+    :returns: DataFrame from gnomAD API
+    :rtype: DataFrame
+    """
+
+    url = 'https://gnomad.broadinstitute.org/api'
+    query = f"""
+    query{{
+      gene(gene_symbol: "{gene_name}", reference_genome: GRCh38) {{
+        variants(dataset: gnomad_r4)
+        {{
+          variant_id
+          chrom
+          pos
+          ref
+          hgvsc
+          hgvsp
+          alt
+          exome {{
+          ac
+          an
+          ac_hom
+            populations
+            {{
+              id
+              ac
+              an
+            }}
+          }}
+          genome
+          {{
+            ac
+            an
+            ac_hom
+            populations
+            {{
+              id
+              ac
+              an
+            }}
+          }}
+        }}
+      }}
+    }}
+    """
+
+    response = requests.post(url, json={'query': query}, timeout=300)# timeout set to 5 minutes
+
+    if response.status_code != 200:
+        print('Error:', response.status_code)
+        return None
+
+    data = response.json()['data']['gene']['variants']
+
+    df = pd.json_normalize(data)
+
+    df['total_ac'] = df['exome.ac'].fillna(0) + df['genome.ac'].fillna(0)
+    df['total_an'] = df['exome.an'].fillna(0) + df['genome.an'].fillna(0)
+
+    df['HGVS Consequence'] = df['hgvsc'].fillna(0) # cDNA change
+    df['Protein Consequence'] = df['hgvsp'].fillna(0) # Protein change
+
+    df['Allele Frequency'] = df['total_ac'] / df['total_an']
+    df['Homozygote Count'] = df['exome.ac_hom'].fillna(0) + df['genome.ac_hom'].fillna(0)
+    exome_populations = df['exome.populations']
+    genome_populations = df['genome.populations']
+    ids = ['afr', 'eas', 'asj', 'sas', 'nfe', 'fin', 'mid', 'amr', 'ami', 'remaining']
+
+    for i in range(len(exome_populations)):
+        exome_pop = exome_populations[i]
+        process_population_data(df, exome_pop, 'exome', ids, i)
+        genome_pop = genome_populations[i]
+        process_population_data(df, genome_pop, 'genome', ids, i)
```

Comment:
Purpose of this block?

Reply:
Calculates and compares all popmaxes and assings popmax population for a variant